command/upgrade/provider.ts

```diff
@@ -21,6 +21,8 @@ export abstract class Provider {
   protected readonly maxListSize: number = 25;
   private maxCols = 8;
 
+  abstract hasRequiredPermissions(): Promise<boolean>;
+
```

Comment:
Adding this method without a default implementation means anyone implementing their own provider will have to implement it.
Would you prefer a default implementation to maintain BC?